test/functional/test_framework/chacha20.py

```diff
@@ -0,0 +1,161 @@
+        """
+        QUARTER_ROUNDS = [
+            # columns
+            (0, 4, 8, 12),
+            (1, 5, 9, 13),
+            (2, 6, 10, 14),
+            (3, 7, 11, 15),
+            # diagonals
+            (0, 5, 10, 15),
+            (1, 6, 11, 12),
+            (2, 7, 8, 13),
+            (3, 4, 9, 14)
+        ]
+
+        for a, b, c, d in QUARTER_ROUNDS:
+            s[a] = (s[a] + s[b]) & 0xffffffff
+            s[d] = ChaCha20.__rot32(s[d] ^ s[a], 16)
+            s[c] = (s[c] + s[d]) & 0xffffffff
+            s[b] = ChaCha20.__rot32(s[b] ^ s[c], 12)
+            s[a] = (s[a] + s[b]) & 0xffffffff
+            s[d] = ChaCha20.__rot32(s[d] ^ s[a], 8)
+            s[c] = (s[c] + s[d]) & 0xffffffff
+            s[b] = ChaCha20.__rot32(s[b] ^ s[c], 7)
+
+    @staticmethod
+    def __rot32(v, bits):
+        """Rotate the 32-bit value v left by bits bits."""
+        bits %= 32  # Make sure the term below does not throw an exception
+        return ((v << bits) & 0xffffffff) | (v >> (32 - bits))
+
+class TestFrameworkChaCha20(unittest.TestCase):
+    def test_cpp(self):
+        def cpp_tester(i, msg_hex, key_hex, nonce_int, counter_int, out_hex):
+            c = chacha20_bindings.ChaCha20()
+            msg_bytes = bytes.fromhex(msg_hex)
+            key_bytes = bytes.fromhex(key_hex)
+            expected_bytes = bytes.fromhex(out_hex)
+
+            c.SetKey(key_bytes)
+            c.SetIV(nonce_int)
+            c.Seek(counter_int)
+
+            if len(msg_hex) == 0:
+                out = c.Keystream(len(expected_bytes))
+            else:
+                out = c.Crypt(msg_bytes)
+
+            self.assertEqual(out, expected_bytes, "ChaCha20 cpp test vector %i" % i)
+
+        self.csv_vectors(cpp_tester)
+
+    def test_python(self):
+        def python_tester(i, msg_hex, key_hex, nonce_int, counter_int, out_hex):
+            msg_bytes = bytes.fromhex(msg_hex)
+            key_bytes = bytes.fromhex(key_hex)
+            expected_bytes = bytes.fromhex(out_hex)
+
+            c = ChaCha20(key_bytes, nonce_int)
+
+            if len(msg_hex) == 0:
+                out = c.keystream(len(expected_bytes), counter_int)
+            else:
+                out = c.encrypt(msg_bytes, counter_int)
+
+            self.assertEqual(out, expected_bytes, "ChaCha20 python test vector %i" % i)
+
+        self.csv_vectors(python_tester)
+
+    def csv_vectors(self, tester):
+        with open(
+            os.path.join(sys.path[0], 'test_framework', 'chacha20_test_vectors.csv'),
+            newline='',
+            encoding='utf8'
+        ) as csvfile:
+            reader = csv.reader(csvfile)
+            next(reader)
+
+            for row in reader:
+                (i_str, msg_hex, key_hex, nonce_str, counter_str, out_hex) = row
+                i = int(i_str)
+                counter_int = int(counter_str)
+                nonce_int = int(nonce_str, 16)
+
+                tester(i, msg_hex, key_hex, nonce_int, counter_int, out_hex)
```

Comment:
since you're always converting `msg_hex`, `key_hex` and `out_hex` to in the tester function, perhaps just do it here?